Refactor Volunteer Leaderboard component to use Apollo hooks and Garden

ThuTrinh opened the pull request:
## Description
- Moved Leaderboard view and data fetch logic into its own component within the Dashboard folder
   - Refactored it to use Apollo hooks
   - Refactored it to use as much Garden components as possible
   - Added loading skeleton views for the user list, avatar and office filter
   - Add Garden style alert for error state
- Moved Milestones view and fetch logic into its own component within the Dashboard folder
   - Refactored it to use Apollo hooks
   - Removed dependency on Redux (by using `withRouter` to get current location)
- Updated the ShowFilter, EventFilter and OfficeFilter to use Garden dropdowns
   - Rearranged the Filters in places to fix alignment and spacing issues (refer to screen shots)
- NamedAvatar
   - Move the confirm delete modal logic out of NamedAvatar and into the UserList itself.
   - Removed dependency on redux
- UserList (Leaderboard & Event details)
   - Refactored to use as much Garden components as possible 
   - Updated the confirm delete modal to use Garden
- ListItem
   - Added a new ListItem component that is a box with a bottom border and when you place 2 things into it, they will be separated by a gap and hug the sides. 

## References
[Github Issue](https://github.com/zendesk/volunteer_portal/issues/164)

## Implementation notes
- Moving the Leaderboard and Milestones out into their own components has the following benefits:
   - Easier to compose and reason about
   - They are self contained and don't affect other components on the page when they fail, since they handle their own data fetching, error handling and loading states.
   - Easier to maintain. They are decoupled from everything else and can be easily moved around or deleted without much effort.
- Switched to using `styled-components` over css-loader.
   - Lets us keep the styles closer to where they are being used.
   - Easily see when styles are not being used during refactoring.  
   - We can leverage prop passing into our styles. This will become more useful when we enable themes properly.
   - This will be the standard for this project going forward.
- Each component should be responsible for its own loading skeleton. 
   - It should support a `loading` prop that determines whether the component should render its skeleton view instead. 
   - This keeps the skeleton loaders close to their real counterparts, enabling them to share any common layout styles. 
   - The `loading` prop gives us an easy interface to trickle down loading state and compose skeleton loaders.
- I chose not to refactor the styling of Milestones too much as I think this component needs a redesign anyway. As it stands, I don't think it is very useful.

## Screenshots
<details>
<summary>Before</summary>
<strong>Loaded State</strong>
<hr/>
<img src="https://user-images.githubusercontent.com/3032977/72615826-a92fb400-3989-11ea-9fba-7c82db3c7ebb.gif">
<strong>Loading State</strong>
<hr/>
<img src="https://user-images.githubusercontent.com/3032977/72615843-ae8cfe80-3989-11ea-8a7c-33756387f124.gif">
<strong>Misaligned Filter
</strong>
<hr/>
<img src="https://user-images.githubusercontent.com/3032977/72662969-4ac60c80-3a41-11ea-8cfc-aa1d89719b00.png">
</details>

<details>
<summary>After</summary>
<strong>Loaded State</strong>
<hr/>
<img src="https://user-images.githubusercontent.com/3032977/72615872-c19fce80-3989-11ea-8e1d-50a36c0e586f.gif">
<strong>Loading State</strong>
<hr/>
<img src="https://user-images.githubusercontent.com/3032977/72615879-c49abf00-3989-11ea-9d18-a64cd7db5375.gif">
<strong>Error State</strong>
<hr/>
<img src="https://user-images.githubusercontent.com/3032977/72662888-5fee6b80-3a40-11ea-8388-cb175dc05b3c.png">
<strong>Aligned/Repositioned Filter</strong>
<hr/>
<img src="https://user-images.githubusercontent.com/3032977/72662983-806af580-3a41-11ea-8076-6deeaf08dea0.png">
<strong>Aligned/Repositioned Filter</strong>
<hr/>
<img src="https://user-images.githubusercontent.com/3032977/72662999-b1e3c100-3a41-11ea-8d26-a8cdd635ff6d.png">
<strong>User List
</strong>
<hr/>
<img src="https://user-images.githubusercontent.com/3032977/72664455-33434f80-3a52-11ea-8ce6-3a939979f7eb.gif">
</details>


## CCs
@zendesk/volunteer

## Risks
* [medium] - there might be some regression because some common components are changed. A full manual test session will need to be done on stg.

ThuTrinh committed:
:construction: Move leaderboard logic into its own component

For better separation of concerns, composability and overall maintenance reasons. Motivated by the need to update it to use apollo hooks and garden.

ThuTrinh committed:
:recycle: Update Leaderboard component to use apollo hooks

ThuTrinh committed:
:recycle: Refactor volunteer leaderboard to use garden and skeleton loaders

Pancrisp commented:
Yessss, skeleton loaders! I was just thinking of swapping out the loaders we currently have for the skeleton ones.

Thanks for kicking things off and getting the ball rolling.

ThuTrinh committed:
:recycle: Update filters to use garden dropdowns

ThuTrinh committed:
:recycle: Fix alignment of filters

After the garden refactor for filters their alignment was a little off on some pages.

ThuTrinh committed:
:recycle: Move user milstones into its own component
- Switch to using Apollo hooks
- Remove dependency on Redux
- Use withRouter to get current location
- Didn't clean up the styling because this component needs to be redesigned. Not sure it is very useful at the moment.

ThuTrinh committed:
:recycle: Move confirm modal out of NamedAvatar component

- Moved the delete confirm modal into the UserList component so that NamedAvatar is more composable.
- Also cleaned up the styling and switch to using the Garden modal component.

ThuTrinh committed:
:art: Wow

samzx commented on app/javascript/components/OfficeFilter/index.js:
```diff
@@ -1,22 +1,53 @@
+      </Select>
+    </Field>
+    <Menu>
+      {[1, 2, 3, 4].map(i => (
```
What's your opinion on `Array(4).fill().map((_, i) => ...` instead? I think you used it that way in `app/javascript/pages/Dashboard/Leaderboard/index.js`

ThuTrinh replied:
I didn't here because 4 is such a small number and I save on an iteration. But I am not really fused. If it's for consistency then I can change it.

samzx commented on app/javascript/components/Toolbar/index.js:
```diff
@@ -39,12 +39,16 @@ const Toolbar = ({ label, view, views, onNavigate, onViewChange, filters, office
-          <ShowFilter value={filters.showFilter.value} onChange={filters.showFilter.onChange} />
-          <EventFilter value={filters.eventFilter.value} onChange={filters.eventFilter.onChange} />
-          <OfficeFilter value={filters.officeFilter.value} onChange={filters.officeFilter.onChange} offices={offices} />
+        <div className="u-display-inline-flex u-mb-xxs">
```
I wonder if it's bad to be mixing the use of both the spacing utility classNames with styled components. I got kind of confused where these class names came from (but then I googled ha).

Is it possible to extract the spacing utilities from https://garden.zendesk.com/css-components/utilities/spacing/ and access them via styled components and use `styled.div` or something along that line?

Maybe not in this PR, but just a thought for the future 🤔

ThuTrinh replied:
I also prefer styled components over the class names. I can access the spacing values through the garden css variables. I will do that instead :)

ThuTrinh replied:
@samzx what are your thought on using something like this? https://styled-system.com
I have used it before and really like it. When done right, you end up with far less components.

samzx replied:
Looks like an interesting concept 😮 I'll have to have a try to see what it is

samzx commented on app/javascript/components/UserList/index.js:
```diff
@@ -1,26 +1,79 @@
-)
+      {activeUser && (
+        <Modal onClose={cancelRemove}>
+          <Header>🤔Remove volunteer from this event? </Header>
```
love this 😹

samzx commented on app/javascript/pages/Dashboard/Leaderboard/index.js:
```diff
@@ -0,0 +1,107 @@
+    },
+  })
+
+  const ofifces = R.propOr([], 'offices', data)
```
offices? 😂

ThuTrinh replied:
woops

samzx commented on app/javascript/pages/Dashboard/Leaderboard/index.js:
```diff
@@ -0,0 +1,107 @@
+
+const Leaderboard = _props => {
+  const currentUser = useContext(UserContext)
+  const [currenOfficeId, setCurrentOfficeId] = useState(R.path(['office', 'Id'], currentUser))
```
currentOfficeId 😛

samzx commented on app/javascript/pages/Dashboard/Leaderboard/index.js:
```diff
@@ -0,0 +1,107 @@
+const Loader = _props => (
+  <>
+    {Array(10)
+      .fill(null)
```
is the `null` necessary? - I think it still works if you map over undefined?

ThuTrinh replied:
yeah it should still work.

samzx reviewed (commented):
Awesome work 🎉 this is really, really great, I'm so happy you did this! 😊

ThuTrinh committed:
:ok_hand: Fix typos

ThuTrinh committed:
:ok_hand: Remove use of className, use styled components instead

Also wired up a basic light theme. Only contains spacing values for now.

ThuTrinh committed:
:white_check_mark: snapshots

ThuTrinh committed:
:recycle: Add no data state for user list and leaderboard

samzx reviewed (approved):
🙌 Very nice ! Think you just [missed](https://github.com/zendesk/volunteer_portal/pull/301/files#diff-14bf74813f38072a8c0ed2bed9e3524bR65) `ofifces`, otherwise LGTM

ThuTrinh committed:
:ok_hand: typo
